File: src/v8/experts/volume_confirmed_breakout.py
"""Volume-gated breakout behavior family (`volume_confirmed_breakout`).

Hypothesis: a close beyond the windowed prior extreme is a breakout setup
ONLY when volume confirms it. Volume is a positive confirmation gate layered
on the breakout trigger (Dow: "volume should be increasing in the direction
of the trend", Ch2.2 p61; Ch6.1 p174-181) — never a stop or target input and
never a standalone entry signal (volume is secondary to price, Ch6.1 p176).

Four book variants share one breakout trigger — close beyond the 20-bar
windowed prior extreme — and differ only in the volume gate:
  a  Dow volume confirmation: breakout-bar volume above its 20-bar smoothed
     average (expanding volume in the breakout direction).
  b  Low-volume breakout timing: volume near its historical minimum
     (vol_min_proximity < 0.4) yet expanding above the smoothed average
     (Ch6.1 p180: "whenever volume declines close to its lowest historical
     level, a breakout is expected" — the increase on the breakout bar is
     the confirmation).
  c  High-volume continuation-bar confirm: breakout-bar volume >= 1.2x the
     smoothed average.
  d  Volume spike, not climax: a >= 2.0x spike that is NOT a 2-sigma
     overextension (vol_zscore < 2.0) — an overextended climax spike belongs
     to the volume_exhaustion family, not to a continuation confirmation.

evaluate() emits ONE CandidateDraft per bar for the highest-priority variant
whose gate fires (declared priority d > c > b > a: most restrictive first, so
every variant owns a non-empty domain and the single-draft dispatch is
deterministic). All four variants count as one multiplicity unit in the
family-level correction (rule 13; D-044 variants_evaluated) and are
distinguished by the `variant` geometry key so episode attribution is exact.
"""
from __future__ import annotations

from ..schema import MarketState, CandidateDraft, ExpertEvaluation
from .base import Expert


class VolumeConfirmedBreakoutExpert(Expert):
    """Close beyond the 20-bar windowed prior extreme, volume-confirmed."""
    expert_id = 'volume_confirmed_breakout'
    version = 'v1'
    mechanism_family_id = 'volume_confirmation'
    behavior_family_id = 'volume_confirmed_breakout'
    variant_id = 'a'
    variants_evaluated = ('a', 'b', 'c', 'd')
    search_universe_size = 4
    requires = ('location', 'volatility', 'participation', 'history')

    # Declared, LOCKED constants (D-036 pattern: declared, never fitted).
    BREAKOUT_WINDOW = 20        # Donchian window (matches window_high_20)
    LOW_VOL_PROXIMITY_MAX = 0.4
    HIGH_VOL_MULT = 1.2
    SPIKE_MULT = 2.0
    CLIMAX_Z = 2.0
    VARIANTS = ('d', 'c', 'b', 'a')     # single-draft gate priority

    def _prior_high(self, i: int) -> float:
        lo = max(0, i - self.BREAKOUT_WINDOW)
        return max(h for (_e, _o, h, _l, _c, _f, _s) in self._hist[lo:i])

    def _prior_low(self, i: int) -> float:
        lo = max(0, i - self.BREAKOUT_WINDOW)
        return min(l for (_e, _o, _h, l, _c, _f, _s) in self._hist[lo:i])

    def _long_pred(self, i: int, bar: tuple) -> bool:
        """Close above the windowed prior high (bounded at the history edge:
        anchors older than the 32-bar window are unstable — documented bound)."""
        if i == 0:
            return False
        _e, _o, _h, _l, close, _f, _s = bar
        return close > self._prior_high(i)

    def _short_pred(self, i: int, bar: tuple) -> bool:
        if i == 0:
            return False
        _e, _o, _h, _l, close, _f, _s = bar
        return close < self._prior_low(i)

    def _evaluate_variants(self, f: dict, sym: str, sma: float) -> str | None:
        """First variant whose volume gate fires, in declared priority order."""
        volume = float(f[f'{sym}.volume'].value)
        for variant in self.VARIANTS:
            if variant == 'd':
                zs = f.get(f'{sym}.vol_zscore')
                if (sma > 0 and volume >= self.SPIKE_MULT * sma
                        and zs is not None and zs.value is not None
                        and float(zs.value) < self.CLIMAX_Z):
                    return variant
            elif variant == 'c':
                if sma > 0 and volume >= self.HIGH_VOL_MULT * sma:
                    return variant
            elif variant == 'b':
                prox = f.get(f'{sym}.vol_min_proximity')
                if (prox is not None and prox.value is not None
                        and float(prox.value) < self.LOW_VOL_PROXIMITY_MAX
                        and sma > 0 and volume > sma):
                    return variant
            else:  # 'a'
                if sma > 0 and volume > sma:
                    return variant
        return None

    def evaluate(self, state: MarketState) -> ExpertEvaluation:
        t = state.as_of
        sym = state.universe[0]
        need = [f'{sym}.close', f'{sym}.atr', f'{sym}.volume',
                f'{sym}.vol_smooth_ma', f'{sym}.history']
        if not self._need(state, need):
            return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                    'NOT_APPLICABLE', 'NO_HABITAT', t)
        f = state.features
        close = float(f[f'{sym}.close'].value)
        atr = f[f'{sym}.atr'].value
        hist_value = f[f'{sym}.history'].value
        if not isinstance(hist_value, (tuple, list)) or not hist_value or atr is None:
            return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                    'NOT_APPLICABLE', 'NO_HABITAT', t)
        self._hist = tuple(hist_value)
        # The breakout levels are the windowed prior extremes of the newest
        # bar (window_high_20 / window_low_20), warmup-gated: a tape without
        # a 20-bar window has no range to break out of.
        wh = f.get(f'{sym}.window_high_20')
        wl = f.get(f'{sym}.window_low_20')
        if wh is None or wh.value is None or wl is None or wl.value is None:
            return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                    'NOT_APPLICABLE', 'NO_HABITAT', t)
        long_level = float(wh.value)
        short_level = float(wl.value)
        if not (close > long_level or close < short_level):
            return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                    'NOT_APPLICABLE', 'NO_SETUP', t)
        variant = self._evaluate_variants(f, sym,
                                          float(f[f'{sym}.vol_smooth_ma'].value))
        if variant is None:
            return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                    'NOT_APPLICABLE', 'NO_SETUP', t)
        if close > long_level:
            direction, level, ref_key = 'LONG', long_level, 'prior_low_ref'
            pred = self._long_pred
        else:
            direction, level, ref_key = 'SHORT', short_level, 'prior_high_ref'
            pred = self._short_pred
        anchor = self.find_setup_anchor(self._hist, pred)
        draft = CandidateDraft(
            expert_id=self.expert_id, expert_version=self.version,
            instrument=sym, direction=direction,
            setup_fingerprint=f'{sym}:{close:.6f}:{level:.6f}:{variant}',
            risk_geometry={'entry': 'NEXT_BAR_CLOSE', 'target_r': 1.0,
                           'stop_r': 1.0, 'expiry_bars': 8,
                           'atr_ref': atr, 'variant': variant,
                           ref_key: level},
            birth_time=t, setup_anchor_event_id=anchor)
        return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                'APPLICABLE', 'CANDIDATE', t, draft)

    def still_valid(self, state: MarketState, draft: CandidateDraft) -> bool:
        """The thesis is "the volume-confirmed breakout holds": price must
        stay on the breakout side of the FROZEN broken level. A close back
        through the level says the breakout did not hold, whatever the stop
        distance still says (Ch4.1 p104: "close back inside the range")."""
        sym = draft.instrument
        f = state.features
        close = f.get(f'{sym}.close')
        if close is None or close.value is None:
            return True          # unobservable: fail open, price still governs
        if draft.direction == 'LONG':
            ref = draft.risk_geometry.get('prior_low_ref')
            if ref is None:
                return True
            return float(close.value) > float(ref)
        ref = draft.risk_geometry.get('prior_high_ref')
        if ref is None:
            return True
        return float(close.value) < float(ref)

File: v8-core/src/experts/volume_confirmed_breakout.rs
//! volume_confirmed_breakout: close beyond the 20-bar window extreme with a
//! volume gate (variants d/c/b/a in priority order); the frozen prior extreme
//! is the reference for gate + anchor. Ported at S4; draft parity proven.

use crate::experts::base::*;
use crate::simulator::Draft;
use crate::state::HistBar;

pub const PORTED: bool = true;
pub const VERSION: &str = "v1";
pub const REQUIRES: &[&str] = &["location", "volatility", "participation", "history"];
// Declared risk geometry (EXPERT_PROTOCOL §1: risk geometry is "Predeclared
// entry, stop, target, timeout and sizing inputs"; SIMULATION_TRUTH_SPEC D-028:
// R is a declared price distance). Fixed values are declared here, never
// re-literalized inside evaluate(); a structural target/stop is computed at
// the call site and overrides the key.
pub const TARGET_R: f64 = 1.0;
pub const STOP_R: f64 = 1.0;
pub const EXPIRY_BARS: i64 = 8;

pub fn volume_confirmed_breakout(fm: &FeatMap, expert_id: &str, version: &str) -> ExpertEval {
    let sym = fm.symbol;
    let close = match fm.value("close") {
        Some(v) => v,
        None => return no_habitat(expert_id, version, fm.as_of),
    };
    let atr = match fm.value("atr") {
        Some(v) => v,
        None => return no_habitat(expert_id, version, fm.as_of),
    };
    let volume = match fm.value("volume") {
        Some(v) => v,
        None => return no_habitat(expert_id, version, fm.as_of),
    };
    let sma = match fm.value("vol_smooth_ma") {
        Some(v) => v,
        None => return no_habitat(expert_id, version, fm.as_of),
    };
    let long_level = match fm.value("window_high_20") {
        Some(v) => v,
        None => return no_habitat(expert_id, version, fm.as_of),
    };
    let short_level = match fm.value("window_low_20") {
        Some(v) => v,
        None => return no_habitat(expert_id, version, fm.as_of),
    };
    if fm.history.is_empty() {
        return no_habitat(expert_id, version, fm.as_of);
    }
    if !(close > long_level || close < short_level) {
        return no_setup(expert_id, version, fm.as_of);
    }
    // _evaluate_variants: first variant whose volume gate fires, in declared
    // priority order d, c, b, a.
    let mut variant: Option<&str> = None;
    if sma > 0.0 && volume >= 2.0 * sma {
        if let Some(z) = fm.value("vol_zscore") {
            if z < 2.0 {
                variant = Some("d");
            }
        }
    }
    if variant.is_none() && sma > 0.0 && volume >= 1.2 * sma {
        variant = Some("c");
    }
    if variant.is_none() {
        if let Some(prox) = fm.value("vol_min_proximity") {
            if prox < 0.4 && sma > 0.0 && volume > sma {
                variant = Some("b");
            }
        }
    }
    if variant.is_none() && sma > 0.0 && volume > sma {
        variant = Some("a");
    }
    let variant = match variant {
        Some(v) => v,
        None => return no_setup(expert_id, version, fm.as_of),
    };
    let (direction, level, ref_key) = if close > long_level {
        ("LONG", long_level, "prior_low_ref")
    } else {
        ("SHORT", short_level, "prior_high_ref")
    };
    let prior_high = |i: usize| -> f64 {
        let lo = i.saturating_sub(20);
        fm.history[lo..i]
            .iter()
            .map(|b| b.high)
            .fold(f64::NEG_INFINITY, f64::max)
    };
    let prior_low = |i: usize| -> f64 {
        let lo = i.saturating_sub(20);
        fm.history[lo..i]
            .iter()
            .map(|b| b.low)
            .fold(f64::INFINITY, f64::min)
    };
    let pred: Box<dyn Fn(usize, &HistBar) -> bool> = if direction == "LONG" {
        Box::new(move |i, b| i > 0 && b.close > prior_high(i))
    } else {
        Box::new(move |i, b| i > 0 && b.close < prior_low(i))
    };
    let anchor = find_setup_anchor(&fm.history, &*pred);
    let draft = Draft {
        direction: direction.to_string(),
        birth_time: fm.as_of,
        risk_geometry: geom(vec![
            ("entry", serde_json::json!("NEXT_BAR_CLOSE")),
            ("target_r", serde_json::json!(TARGET_R)),
            ("stop_r", serde_json::json!(STOP_R)),
            ("expiry_bars", serde_json::json!(EXPIRY_BARS)),
            ("atr_ref", serde_json::json!(atr)),
            ("variant", serde_json::json!(variant)),
            (ref_key, serde_json::json!(level)),
        ]),
    };
    let fingerprint = format!("{sym}:{:.6}:{:.6}:{}", close, level, variant);
    candidate(expert_id, version, fm.as_of, draft, anchor, fingerprint)
}
